### PR TITLE
Client cookie refactoring

### PR DESCRIPTION
**Pull request by LavrovArtem**
related to #1137

### CODE REVIEW

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

File: src/client/sandbox/cookie/index.js [hunk 1]
```diff
@@ -139,9 +106,8 @@ export default class CookieSandbox extends SandboxBase {
     }
 
     setCookie (document, cookie, syncWithServer) {
-        // eslint-disable-next-line no-restricted-properties
-        if (cookie.length > BYTES_PER_COOKIE_LIMIT || destLocation.getParsed().protocol === 'file:')
-            return cookie;
+        if (this._isCannotSetCookie())
```
Comment (miherlosev) [inline src/client/sandbox/cookie/index.js]: Add comment for this line about covered behavior.

We usually don't use function names with negation.
So rewrite as
```js
if (!this._isCanSetCookie())
    return;
```

File: src/client/sandbox/cookie/index.js [hunk 2]
```diff
@@ -26,78 +28,43 @@ export default class CookieSandbox extends SandboxBase {
...
-    }
-
     // NOTE: Perform validations that can't be processed by a browser due to proxying.
     static _isValidCookie (parsedCookie) {
         if (!parsedCookie)
             return false;
 
         // NOTE: HttpOnly cookies can't be accessed from the client code.
-        if (parsedCookie.httponly)
+        if (parsedCookie.httpOnly)
```
[the remark this answers is not in the document]
Comment (miherlosev) [reply src/client/sandbox/cookie/index.js]: Ok

File: src/client/sandbox/cookie/index.js [hunk 3]
```diff
@@ -26,78 +28,43 @@ export default class CookieSandbox extends SandboxBase {
...
+        const clientCookie = `key${nativeMethods.random.call(Math)}=value`;
+
+        nativeMethods.documentCookieSetter.call(this.document, clientCookie);
+
+        const documentCookieIsEmpty = !nativeMethods.documentCookieGetter.call(this.document);
+
+        if (!documentCookieIsEmpty)
+            nativeMethods.documentCookieSetter.call(this.document, `${clientCookie};expires=${MIN_DATE_VALUE}`);
+
+        return documentCookieIsEmpty;
```
Comment (churkin) [inline src/client/sandbox/cookie/index.js]: As I understand you can not set cookie if the `documentCookieIsEmpty` is `true`? Where am I wrong?

File: src/client/sandbox/cookie/index.js [hunk 4]
```diff
@@ -151,53 +118,29 @@ export default class CookieSandbox extends SandboxBase {
         if (setByClient) {
             this.syncCookie();
 
+            // eslint-disable-next-line no-restricted-properties
+            if (cookie.length > BYTES_PER_COOKIE_LIMIT || destLocation.getParsed().protocol === 'file:')
```
Comment (churkin) [inline src/client/sandbox/cookie/index.js]: Move this to the `_canSetCookie(cookie)`

File: src/client/sandbox/native-methods.js [hunk 5]
```diff
@@ -229,9 +229,12 @@ class NativeMethods {
         this.formDataAppend         = win.FormData.prototype.append;
 
         // DateTime
-        this.date    = win.Date;
+        this.Date    = win.Date;
         this.dateNow = win.Date.now;
 
+        // Math
+        this.random = Math.random;
```
Comment (churkin) [inline src/client/sandbox/native-methods.js]: I think we should save `Math`

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

File: src/client/sandbox/cookie/index.js [hunk 6]
```diff
@@ -64,7 +68,7 @@ export default class CookieSandbox extends SandboxBase {
         if (!documentCookieIsEmpty)
             nativeMethods.documentCookieSetter.call(this.document, `${clientCookie};expires=${MIN_DATE_VALUE}`);
 
-        return documentCookieIsEmpty;
+        return !documentCookieIsEmpty;
```
Comment (churkin) [inline src/client/sandbox/cookie/index.js]: Wow. Your function worked conversely but tests were passed. May be we should add tests for this?

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Comment by LavrovArtem** [comment]
ping @churkin

**Pull request merged**